### PR TITLE
Fixed AdminType delete checkboxes

### LINKED ISSUE #4348: Sonata type collection delete row triggers constraints validation
I have a problem with sonata admin, specifically in a type collection field. Here's the scenario:

I have an entity called "Group" with a one to many relationship to the entity "Member"

In sonata admin I set up the admin classes for the 2 entities.

In the "Group admin" i create a collection type with the Member entity this way

```php
->add('members', 'sonata_type_collection', [
    'by_reference'          => false,
    'label'                 => 'Members',
    'type_options'          => ['delete' => true],
    'btn_add'               => "Add Member",
    'required'              => false,
    'constraints'           => $validation['members'],
],
[
    'edit'              => 'inline',
    'inline'            => 'table'
])
```
The member entity has 4 fields, one of which is required.

```php
->add('firstName', TextType::class, [
    'label'         => 'First Name'
    'constraints'   => [
        new NotBlank(['message' => 'Please enter the name.']),
    ]
])
```
If I click the "Add Member" button in the edit view, it adds a new row as expected. At that point, if I change my mind and decide to delete the newly added record, without adding the name, on save it's returning a validation error telling me 'Please enter the name.' on the field.

Shouldn't the delete action have priority over validation?

### PR DESCRIPTION
**Pull request by RomainSanchez**
I am targeting this branch, because this is BC.

Closes #4348 

## Changelog

```markdown

### Changed
When checking the delete checkbox of an inline child form of `CollectionType` the related fields are now disabled to avoid preventing submission of the form when one of those inputs is required.

```

## Subject

See #4348

### CODE REVIEW

**Comment by RomainSanchez** [comment]
So should i move the javascript to Admin.js ?

**Comment by core23** [comment]
Yes please

File: Resources/public/inlineDeleteCheckboxes.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+ */
+
+var handleInlineDeleteCheckboxes = function() {
+    var checkboxes = jQuery('.sonata-admin-type-delete-checkbox');
+
+    if(checkboxes.length) {
+        var eventType = window.SONATA_CONFIG.USE_ICHECK ? 'ifChanged': 'change';
+
+        checkboxes.on(eventType, function() {
```
Comment (jordisala1991) [inline Resources/public/inlineDeleteCheckboxes.js]: Isn't it better to just bind this event to the document with a selector?

Ex: 

`jQuery(document).on(eventType, '.sonata-admin-type-delete-checkbox', function() {`

With this you don't need:

```
jQuery(document).on(
    'ready sonata-admin-setup-list-modal sonata-admin-append-form-element sonata.add_element',
    handleInlineDeleteCheckboxes
);
```

Comment (RomainSanchez) [inline Resources/public/inlineDeleteCheckboxes.js]: `ifchanged, change` these are the events to handle the checkbox being checked
`ready sonata-admin-setup-list-modal sonata-admin-append-form-element sonata.add_element` these ones are to reinitialize the listener to the above events on dynamically added content .Don't really see how i can mix those

**Comment by RomainSanchez** [comment]
I'll add the second commit back in once the first one is done being amended

File: Resources/public/Admin.js [hunk 2]
```diff
@@ -620,10 +636,14 @@ jQuery(document).ready(function() {
...
 });
 
 jQuery(document).on('sonata-admin-append-form-element', function(e) {
     Admin.setup_select2(e.target);
     Admin.setup_icheck(e.target);
     Admin.setup_collection_counter(e.target);
+    Admin.handleInlineDeleteCheckboxes();
 });
+
+jQuery(document).on('sonata.add_element',Admin.handleInlineDeleteCheckboxes);
```
Comment (core23) [inline Resources/public/Admin.js]: Please add a space after the comma

File: Resources/public/Admin.js [hunk 3]
```diff
@@ -608,6 +608,22 @@ var Admin = {
...
+    handleInlineDeleteCheckboxes: function() {
+        var checkboxes = jQuery('.sonata-admin-type-delete-checkbox');
+
+        if(checkboxes.length) {
+            var eventType = window.SONATA_CONFIG.USE_ICHECK ? 'ifChanged': 'change';
+
+            checkboxes.on(eventType, function() {
+                var id = jQuery(this).prop('id');
+
+                jQuery('[id^=' + id.split('__')[0] + ']')
```
Comment (core23) [inline Resources/public/Admin.js]: Can be writte as: `jQuery('[id^=' + id.split('__')[0] + ']:not('#' + id + ')')`

File: Resources/public/inlineDeleteCheckboxes.js (hunk 1 above)
Comment (jordisala1991) [inline Resources/public/inlineDeleteCheckboxes.js]: If you bind an event to document with a selector you don't need to reinitialize the listener on dinamically added content.

**Comment by RomainSanchez** [comment]
Is it fine by you @core23 ?

File: Form/Type/AdminType.php [hunk 4]
```diff
@@ -46,7 +46,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
             }
 
-            $builder->add('_delete', $options['delete_options']['type'], $options['delete_options']['type_options']);
+            $builder->add(
+                '_delete',
+                $options['delete_options']['type'],
+                array_merge(
+                    $options['delete_options']['type_options'],
+                    array('attr' => array(
+                        'class' => 'sonata-admin-type-delete-checkbox', ),
```
Comment (OskarStark) [inline Form/Type/AdminType.php]: the end of the line looks weird

File: Resources/public/Admin.js [hunk 5]
```diff
@@ -620,10 +635,14 @@ jQuery(document).ready(function() {
     Admin.setup_per_page_switcher(document);
     Admin.setup_collection_buttons(document);
     Admin.shared_setup(document);
+    Admin.handleInlineDeleteCheckboxes();
```
Comment (OskarStark) [inline Resources/public/Admin.js]: we should use underscore for the method name

File: Form/Type/AdminType.php (hunk 4 above)
Comment (RomainSanchez) [inline Form/Type/AdminType.php]: I know right, styleCI forced this syntax(apparently it requires all arrays to have a comma after the last item)

Comment (OskarStark) [inline Form/Type/AdminType.php]: the comma is ok, but not the closing bracket

Comment (OskarStark) [inline Form/Type/AdminType.php]: @SpacePossum, this looks weird, what do you think?

Comment (SpacePossum) [inline Form/Type/AdminType.php]: Hi,
My guess its done by `trailing_comma_in_multiline_array` rule.
Looking at the code there are 2 multi line arrays, both starting on line 54, one ending on 55, other on 56. Therefor the fixer will add ',' at the end of each element. Guess this is what it makes look odd. 
You could, for example, update the current;

```php
array('attr' => array(
    'class' => 'sonata-admin-type-delete-checkbox', ),
)
```
to:

```php
array(
    'attr' => array( 'class' => 'sonata-admin-type-delete-checkbox' )
)
```

but you can define the style yourselves of course :)
As I see it, its not a bug in the fixer, it just does what you asked it to do.

Comment (OskarStark) [inline Form/Type/AdminType.php]: Thanks for your Feedback

Comment (RomainSanchez) [inline Form/Type/AdminType.php]: So what's the course of action ?

File: Resources/public/Admin.js [hunk 6]
```diff
@@ -608,6 +608,21 @@ var Admin = {
...
             });
         });
+    },
+    handle_inline_delete_checkboxes: function() {
+        var checkboxes = jQuery('.sonata-admin-type-delete-checkbox');
+
+        if(checkboxes.length) {
+            var eventType = window.SONATA_CONFIG.USE_ICHECK ? 'ifChanged': 'change';
+
+            checkboxes.on(eventType, function() {
```
Comment (jordisala1991) [inline Resources/public/Admin.js]: http://stackoverflow.com/questions/12824549/should-all-jquery-events-be-bound-to-document

Still thinking that it is better to bind it to document. Reasons explained on that link. 

WDYT? @sonata-project/contributors

Comment (greg0ire) [inline Resources/public/Admin.js]: I'm unsure. Is the goal to be sure not to miss any checkbox created w/ AJAX?

Comment (RomainSanchez) [inline Resources/public/Admin.js]: http://stackoverflow.com/a/12824698/5758328

@greg0ire yep

Comment (greg0ire) [inline Resources/public/Admin.js]: Maybe bind sth smaller, like the main form?

File: Form/Type/AdminType.php (hunk 4 above)
Comment (jordisala1991) [inline Form/Type/AdminType.php]: This can all be on the same line, it is not even 80 chars

File: Form/Type/AdminType.php [hunk 7]
```diff
@@ -46,7 +46,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 $options['delete_options']['type_options']['translation_domain'] = $admin->getTranslationDomain();
             }
 
-            $builder->add('_delete', $options['delete_options']['type'], $options['delete_options']['type_options']);
+            $builder->add(
+                '_delete',
+                $options['delete_options']['type'],
+                array_merge(
+                    $options['delete_options']['type_options'],
+                    array('attr' => array(
```
Comment (core23) [inline Form/Type/AdminType.php]: This looks wrong too. Please break this line, so each array key as a newline:


```
          $builder->add(
                '_delete',
                $options['delete_options']['type'],
                array_merge(
                    $options['delete_options']['type_options'],
                    array(
                          'attr' => array(
```

or Maybe
```
          $builder->add(
                '_delete',
                $options['delete_options']['type'],
                array_merge($options['delete_options']['type_options'], array(
                    'attr' => array(
```

**Comment by RomainSanchez** [comment]
All good @core23 @OskarStark ?

**Comment by jordisala1991** [comment]
@OskarStark @greg0ire please review

**Comment by OskarStark** [comment]
Thank you @RomainSanchez 👍  for the great contribution 🎉

**Comment by jordisala1991** [comment]
This does not work when the delete checkbox is disabling an item with validation.
The old code was working on that case...

Use case:
Edit an entity with a collection and some items already added. Items have a validation on the text (not null). Check one delete checkbox. Save. -> Error because it does not pass validation (it should not be disabled).

Ping @RomainSanchez @OskarStark @greg0ire

**Comment by RomainSanchez** [comment]
Didn't manage to reproduce your case.I found another problem though: 

If you check the delete checkbox for one of the existing items and then add a new item since the fields of the existing one are disabled when the form is sent via ajax it returns empty fields.

Maybe items should just be removed via ajax when checking the delete checkbox since they are added via ajax.

**Comment by jordisala1991** [comment]
I will work on an example to ilustrate the problem. But IMO it doesn't look easy to fix, maybe we need to revert this one...

Will try to create the example tonight

**Comment by cosf87** [comment]
I get an error after this update. For instance, I have three DB tables:

**Activity** ---_OneToMany_--- **ActivityAdaptability** ---_ManyToOne_--- **Adaptability**

ActivityAdaptability is an inline CollectionType field in ActivityAdmin
```
->add('activityAdaptabilities', CollectionType::class,
                        array(
                            'by_reference' => false,
                            'label' => false,
                        ),
                        array(
                            'admin_code' => 'sonata.admin.activity_adaptability',
                            'edit' => 'inline',
                            'inline' => 'table',
                            'sortable' => 'position'
                        )
                    )
```
When I try to delete an ActivityAdaptability item I get this stack trace:
```
[1] Symfony\Component\PropertyAccess\Exception\InvalidArgumentException: Expected argument of type "AppBundle\Entity\Adaptability", "NULL" given
    at n/a
        in vendor/symfony/symfony/src/Symfony/Component/PropertyAccess/PropertyAccessor.php line 254
    at Symfony\Component\PropertyAccess\PropertyAccessor::throwInvalidArgumentException('Argument 1 passed to AppBundle\Entity\ActivityAdaptability::setAdaptability() must be an instance of AppBundle\Entity\Adaptability, null given, called in vendor/symfony/symfony/src/Symfony/Component/PropertyAccess/PropertyAccessor.php on line 605', array(array('file' => 'vendor/symfony/symfony/src/Symfony/Component/PropertyAccess/PropertyAccessor.php', 'line' => '605', 'function' => 'setAdaptability', 'class' => 'AppBundle\Entity\ActivityAdaptability', 'type' => '->', 'args' => array(null)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/PropertyAccess/PropertyAccessor.php', 'line' => '201', 'function' => 'writeProperty', 'class' => 'Symfony\Component\PropertyAccess\PropertyAccessor', 'type' => '->', 'args' => array(array(object(ActivityAdaptability), object(ActivityAdaptability)), 'adaptability', null)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php', 'line' => '93', 'function' => 'setValue', 'class' => 'Symfony\Component\PropertyAccess\PropertyAccessor', 'type' => '->', 'args' => array(object(ActivityAdaptability), object(PropertyPath), null)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/Form/Form.php', 'line' => '623', 'function' => 'mapFormsToData', 'class' => 'Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper', 'type' => '->', 'args' => array(object(RecursiveIteratorIterator), object(ActivityAdaptability))), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/Form/Form.php', 'line' => '567', 'function' => 'submit', 'class' => 'Symfony\Component\Form\Form', 'type' => '->', 'args' => array(array(), true)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/Form/Form.php', 'line' => '567', 'function' => 'submit', 'class' => 'Symfony\Component\Form\Form', 'type' => '->', 'args' => array(array(array('_delete' => '1'), array('adaptability' => '6', 'position' => '4')), true)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/Form/Extension/HttpFoundation/HttpFoundationRequestHandler.php', 'line' => '113', 'function' => 'submit', 'class' => 'Symfony\Component\Form\Form', 'type' => '->', 'args' => array(array('activityAdaptabilities' => array(array('adaptability' => '3', 'position' => '1'), array('adaptability' => '5', 'position' => '2'), array('_delete' => '1'), array('adaptability' => '6', 'position' => '4'))), true)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/Form/Form.php', 'line' => '489', 'function' => 'handleRequest', 'class' => 'Symfony\Component\Form\Extension\HttpFoundation\HttpFoundationRequestHandler', 'type' => '->', 'args' => array(object(Form), object(Request))), array('file' => 'vendor/sonata-project/admin-bundle/Controller/CRUDController.php', 'line' => '257', 'function' => 'handleRequest', 'class' => 'Symfony\Component\Form\Form', 'type' => '->', 'args' => array(object(Request))), array('function' => 'editAction', 'class' => 'Sonata\AdminBundle\Controller\CRUDController', 'type' => '->', 'args' => array('2')), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php', 'line' => '144', 'function' => 'call_user_func_array', 'args' => array(array(object(CRUDController), 'editAction'), array('2'))), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php', 'line' => '64', 'function' => 'handleRaw', 'class' => 'Symfony\Component\HttpKernel\HttpKernel', 'type' => '->', 'args' => array(object(Request), '1')), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/HttpKernel/DependencyInjection/ContainerAwareHttpKernel.php', 'line' => '69', 'function' => 'handle', 'class' => 'Symfony\Component\HttpKernel\HttpKernel', 'type' => '->', 'args' => array(object(Request), '1', true)), array('file' => 'vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php', 'line' => '185', 'function' => 'handle', 'class' => 'Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel', 'type' => '->', 'args' => array(object(Request), '1', true)), array('file' => 'web/app_dev.php', 'line' => '30', 'function' => 'handle', 'class' => 'Symfony\Component\HttpKernel\Kernel', 'type' => '->', 'args' => array(object(Request)))), '0')
        in vendor/symfony/symfony/src/Symfony/Component/PropertyAccess/PropertyAccessor.php line 218

    at Symfony\Component\PropertyAccess\PropertyAccessor->setValue(object(ActivityAdaptability), object(PropertyPath), null)
        in vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php line 93

    at Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper->mapFormsToData(object(RecursiveIteratorIterator), object(ActivityAdaptability))
        in vendor/symfony/symfony/src/Symfony/Component/Form/Form.php line 623

    at Symfony\Component\Form\Form->submit(array(), true)
        in vendor/symfony/symfony/src/Symfony/Component/Form/Form.php line 567

    at Symfony\Component\Form\Form->submit(array(array('_delete' => '1'), array('adaptability' => '6', 'position' => '4')), true)
        in vendor/symfony/symfony/src/Symfony/Component/Form/Form.php line 567

    at Symfony\Component\Form\Form->submit(array('activityAdaptabilities' => array(array('adaptability' => '3', 'position' => '1'), array('adaptability' => '5', 'position' => '2'), array('_delete' => '1'), array('adaptability' => '6', 'position' => '4'))), true)
        in vendor/symfony/symfony/src/Symfony/Component/Form/Extension/HttpFoundation/HttpFoundationRequestHandler.php line 113
```

**Comment by jordisala1991** [comment]
https://github.com/jordisala1991/ArchetypeSonata

The example is located under Sonata\BaseBundle bundle: https://github.com/jordisala1991/ArchetypeSonata/tree/master/src/Sonata/BaseBundle

It is runnable using docker or vagrant or your custom solution.

Here is the example. When you have not nullable fields on the Collection, and you try to remove one of them with the checkbox. It throws an exception because is trying to set them to NULL (I also had to remove typehinting `string` because is calling `set*(null)`).

On this example it only fails on delete, but I got another one using translations inside Collection that it fails on another case (deleting a non saved entry)

**Comment by RomainSanchez** [comment]
thanks the project works great i managed to reproduce your case

**Comment by jordisala1991** [comment]
So... how do we handle this? I am not sure we can fix this issue without reverting this PR... @greg0ire any ideas?

**Comment by greg0ire** [comment]
Let's put a deadline… if this is not fixed by next monday morning, then revert it?

**Comment by RomainSanchez** [comment]
So revert it is. Couldn't find an easy fix.

https://github.com/sonata-project/SonataAdminBundle/pull/4358/commits/5b31649a682341ecac5cf78bde730dbbbf9e986b might not need to be reverted though

**Comment by greg0ire** [comment]
This was squash merged, but I managed to cherry-pick your commit.

**Pull request merged**